### ISSUE TITLE
File attachment support in the Help Center API

### ISSUE THREAD

**Issue by KenBilling** [mozts2005/ZendeskApi_v2#136]
Currently, if you upload an article that contains links to images, the links are broken when the HC is viewed. You must edit the articles manually in Admin mode and upload the files individually for each image in each article.

That's very tedious and time consuming the first time around. What's worse is the article editor doesn't understand the concept of an image insertion point so even if you position the cursor where you want the image, it will be inserted at the very beginning of the article. So you must delete it from there and reinsert it where you wanted it in the first place.

But wait, there's more!

If you later re-upload your article source to the HC because you've made changes to the text, guess what? Yep, your tediously placed image links are all overwritten and you have to go through the whole process again.

So you see, attachment support in the API is sorely needed unless you don't use any icons or screenshots in your articles and who does that?

**Comment by mozts2005** [comment]
Will look in to this. But it will be faster if you are able to submit a pull request with a failing test case or the fix.

**Comment by KenBilling** [comment]
Sorry I wasn't more clear. There are no test cases to fail because they haven't been written yet because this whole feature is unsupported by the API. The API is incomplete and if I could fix it, I would. Then I wouldn't request support for the feature be added by those contributors who are already working on the API.

**Comment by mozts2005** [comment]
to be clear the API that is provided by zendesk does have the ability to do what you are what you are asking for.  [zendesk api]( https://developer.zendesk.com/rest_api/docs/help_center/articles#associate-attachments-in-bulk-to-article)

However,  we do not support that at this time. but here is the cool thing about open source your are able to add it then commit it back for others to use. 

I am asking for test cases or a pull request as you are the person that will know if the issue has been fix for you. 

when I say test case it may only be a stub creating the objects that you want to be posted correctly. that way we can make sure that any work done will be what you need.

**Comment by mozts2005** [comment]
I fell that the commit 60711245f780c3ee3b5f9a349528bea3c9f246e5 by @comagnan fixed your issue. If it did not I will reopen the issue just let us know.

**Comment by KenBilling** [comment]
Hi, Elizabeth

I updated my project to the latest build to try the new article attachment support and now none of my API calls work. They all return nothing and throw exceptions in my error handler. I’m using build 3.0.29.0 with Newtonsoft.Json 7.01 and authenticating with a Zendesk API token, not a password. Am I missing something (not so) obvious?

Is there somewhere online to discuss working with the wrapper with other programmers?

Thanks,
Ken

**Comment by comagnan** [comment]
What kind of exceptions are we talking about, 401 - Unauthorized?

Before the attachment support changeset (which is only for getting attachments, not setting, so I suppose it's only partially helpful for your situation anyway), I added OAuth token support on top of API tokens for the purpose of the project on which I'm working. This might be the source of the issue.

I am inexperienced with GitHub, so I was taken aback to see my changesets added to the main branch; They weren't tested outside of my specific setup thus I figured it was safe to keep them on a separate fork, 

Sorry for any problems my changeset might have caused you.

**Comment by KenBilling** [comment]
Apparently, that was indeed the problem.

After updating to 3.0.29, I had to remove my passes of Nothing as the password parameter during API object creation. Then I noticed the recent history of build failures of the wrapper and downgraded to 3.0.21, which after I restored the password parameters to the calls, now succeeds again.

But I can’t tell in which build your attachment changes were merged or if they have been yet. I’m still trying to make sense out of GitHub myself.

Seems like the wrapper’s in considerable flux right now and if you’re changeset only supports getting attachments, I’ll continue with building attachment upload support in my own project. It’s going to be complicated anyway since attachments are assigned new IDs after they are uploaded. So I’ll have to upload the attachments before the article content, then patch the article with the IDs, and upload the article content last. I’ll have to store the IDs as hidden metadata in the source system in order to maintain local link resolution. Ugh.

Oh well. If it was easy, anybody could do it, right?

Thanks for the quick reply!
